packages/medusa/src/loaders/knex-connection.ts

```diff
@@ -0,0 +1,50 @@
+import { asValue, AwilixContainer } from "awilix"
+import { DataSourceOptions } from "typeorm"
+import { ConfigModule } from "../types/global"
+import { ContainerRegistrationKeys } from "@medusajs/utils"
+import { knex } from "knex"
+
+type Options = {
+  configModule: ConfigModule
+  container: AwilixContainer
+  customOptions?: {
+    migrations: DataSourceOptions["migrations"]
+    logging: DataSourceOptions["logging"]
+  }
+}
+
+export default async ({ container, configModule }: Options): Promise<void> => {
+  const connectionString = configModule.projectConfig.database_url
+  const database = configModule.projectConfig.database_database
+  const extra: any = configModule.projectConfig.database_extra || {}
+  const schema = configModule.projectConfig.database_schema || "public"
+
+  // Share a knex connection to be consumed by the shared modules
+  if (
+    !container.hasRegistration(
+      ContainerRegistrationKeys.PG_KNEX_CONNECTION_REGISTRATION_KEY
+    )
+  ) {
+    const pgConnection = knex<any, any>({
```

Comment:
thought: Discussed with Seb this morning, whether it would make sense to bring this to an even lower level i.e. establish a Postgres connection using pg (or a similar library). It would remove all potential limitations, that a knex connection could add to the implementation of modules. What do you think? Do you what the knex wrapper gives us, that we wouldn't get from initializing a connection via a pg library?

Reply:
I wanted to do that, but then it is unusable by an orm. Depending on the orm it will accept something specific, knex is just a query builder and the lower level mikro orm can accept. If you look at the AbstractSqlConnection class, it checks that it is an instance of knex.Client. I will look if i can do something like initialising a pg connection and pass it to knex to be then used by mikro orm. But unfortunately we have to compose with the tech we are using in some ways

Reply:
> i think i find a way, i can create a connection pool as our lower level connection. This will be shared to the modules. The modules can then create a knex instance using that connection which can then be passed to mikro orm to be accepted and consumed

Sounds like a good approach!